lib/ui/painting.dart

```diff
@@ -1379,13 +1379,24 @@ class Paint {
   ///
   ///  * [MaskFilter], which is used for drawing geometry.
   ImageFilter get imageFilter {
-    if (_objects == null)
+    if (_objects == null || _objects[_kImageFilterIndex] == null)
       return null;
-    return _objects[_kImageFilterIndex];
+    return _objects[_kImageFilterIndex].creator;
   }
   set imageFilter(ImageFilter value) {
-    _objects ??= List<dynamic>(_kObjectCount);
-    _objects[_kImageFilterIndex] = value;
+    final _ImageFilter nativeFilter = value?._toNativeImageFilter();
+    if (nativeFilter == null) {
+      if (_objects != null) {
+        _objects[_kImageFilterIndex] = null;
+      }
+    } else {
+      if (_objects == null) {
+        _objects = List<dynamic>(_kObjectCount);
+        _objects[_kImageFilterIndex] = nativeFilter;
+      } else if (_objects[_kImageFilterIndex]?.creator != value) {
+        _objects[_kImageFilterIndex] = nativeFilter;
+      }
```

Comment:
I refer back to "I'm following the pattern in ColorFilter even though it has flaws", but to look further into the points you bring up:

- The new code copied from ColorFilter prevents creating an array for storing null into an already null property.
- The looks like it wants to avoid extra work when the value is the same as that which created the old filter, but it doesn't do a good job of that. The first part of the code should test value for null and then it avoids a toNative call, the second part could avoid a toNative call if it would do the test before it already called it, but it doesn't do that.

So, yes, it's a mess, but with a purpose. Should I fix this up and fix ColorFilter to match? Or go with the existing code and file an issue to clean it up in a separate PR?

Reply:
I made some simplifications to this code that will appear in the next commit.